pkg/generator/cloudsmith/cloudsmith.go

```diff
@@ -0,0 +1,203 @@
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cloudsmith
+
+import (
+	"bytes"
+	"context"
+	b64 "encoding/base64"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"net/http"
+	"strings"
+	"time"
+
+	apiextensions "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
+	"k8s.io/klog/v2"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/yaml"
+
+	genv1alpha1 "github.com/external-secrets/external-secrets/apis/generators/v1alpha1"
+	"github.com/external-secrets/external-secrets/pkg/utils"
+)
+
+type Generator struct {
+	httpClient *http.Client
+}
+
+type CloudsmithOIDCRequest struct {
+	OIDCToken   string `json:"oidc_token"`
+	ServiceSlug string `json:"service_slug"`
+}
+
+type CloudsmithOIDCResponse struct {
+	Token string `json:"token"`
+}
+
+const (
+	defaultCloudsmithAPIHost = "api.cloudsmith.io"
+
+	errNoSpec            = "no config spec provided"
+	errParseSpec         = "unable to parse spec: %w"
+	errGetToken          = "unable to get authorization token: %w"
+	errExchangeToken     = "unable to exchange OIDC token: %w"
+	errMarshalRequest    = "failed to marshal request payload: %w"
+	errCreateRequest     = "failed to create HTTP request: %w"
+	errUnexpectedStatus  = "request failed due to unexpected status: %s"
+	errReadResponse      = "failed to read response body: %w"
+	errUnmarshalResponse = "failed to unmarshal response: %w"
+	errTokenNotFound     = "token not found in response"
+	errTokenTypeCast     = "error when typecasting token to string"
+
+	httpClientTimeout = 30 * time.Second
+)
+
+func (g *Generator) Generate(ctx context.Context, cloudsmithSpec *apiextensions.JSON, kubeClient client.Client, targetNamespace string) (map[string][]byte, genv1alpha1.GeneratorProviderState, error) {
+	return g.generate(
+		ctx,
+		cloudsmithSpec,
+		kubeClient,
+		targetNamespace,
+	)
+}
+
+func (g *Generator) Cleanup(_ context.Context, cloudsmithSpec *apiextensions.JSON, providerState genv1alpha1.GeneratorProviderState, _ client.Client, _ string) error {
+	return nil
+}
+
+func (g *Generator) generate(
+	ctx context.Context,
+	cloudsmithSpec *apiextensions.JSON,
+	_ client.Client,
+	targetNamespace string) (map[string][]byte, genv1alpha1.GeneratorProviderState, error) {
+	if cloudsmithSpec == nil {
+		return nil, nil, errors.New(errNoSpec)
+	}
+	res, err := parseSpec(cloudsmithSpec.Raw)
+	if err != nil {
+		return nil, nil, fmt.Errorf(errParseSpec, err)
+	}
+
+	// Fetch the service account token
+	oidcToken, err := utils.FetchServiceAccountToken(ctx, res.Spec.ServiceAccountRef, targetNamespace)
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to fetch service account token: %w", err)
+	}
+
+	apiHost := res.Spec.APIHost
+	if apiHost == "" {
+		apiHost = defaultCloudsmithAPIHost
+	}
+	apiHost = strings.TrimPrefix(apiHost, "https://")
```

Comment:
I would prefer the more robust way of parsing it as a url and removing the scheme. Incidentally, that will be a good validation to see if it's a valid host path. :)